utils/types.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2021 Pinterest
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+
+	"github.com/pinterest/thriftcheck"
+	"go.uber.org/thriftrw/ast"
+)
+
+// TypeMatcher represents a way to match against AST types.
+type TypeMatcher interface {
+	Matches(c *thriftcheck.C, t ast.Type) bool
+	Name() string
+}
+
+// directTypeMatcher matches direct AST types (map, list, set).
+type directTypeMatcher struct {
+	name    string
+	matchFn func(ast.Type) bool
+}
+
+func (m *directTypeMatcher) Matches(c *thriftcheck.C, t ast.Type) bool {
+	// Resolve TypeReference if needed.
+	if typeRef, ok := t.(ast.TypeReference); ok {
+		if resolved := c.ResolveType(typeRef); resolved != nil {
+			if resolvedType, ok := resolved.(ast.Type); ok {
+				t = resolvedType
+			} else {
+				return false
+			}
+		} else {
+			return false // Unresolved type, can't match.
+		}
+	}
+	return m.matchFn(t)
+}
+
+func (m *directTypeMatcher) Name() string {
+	return m.name
+}
+
+// primitiveTypeMatcher matches specific primitive types by BaseTypeID.
+type primitiveTypeMatcher struct {
+	name   string
+	typeID ast.BaseTypeID
+}
+
+func (m *primitiveTypeMatcher) Matches(c *thriftcheck.C, t ast.Type) bool {
+	if typeRef, ok := t.(ast.TypeReference); ok {
+		if resolved := c.ResolveType(typeRef); resolved != nil {
+			if resolvedType, ok := resolved.(ast.Type); ok {
+				t = resolvedType
+			} else {
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+	if baseType, ok := t.(ast.BaseType); ok {
+		return baseType.ID == m.typeID
+	}
+	return false
+}
+
+func (m *primitiveTypeMatcher) Name() string {
+	return m.name
+}
+
+// structureTypeMatcher matches struct-like types (union, struct, exception).
+type structureTypeMatcher struct {
+	name       string
+	structType ast.StructureType
+}
+
+func (m *structureTypeMatcher) Matches(c *thriftcheck.C, t ast.Type) bool {
+	// Struct/union/exception types are always referenced via TypeReference.
+	typeRef, ok := t.(ast.TypeReference)
+	if !ok {
+		return false
+	}
+	resolved := c.ResolveType(typeRef)
+	if resolved == nil {
+		return false
+	}
+	if structDef, ok := resolved.(*ast.Struct); ok {
+		return structDef.Type == m.structType
+	}
+
+	return false
+}
+
+func (m *structureTypeMatcher) Name() string {
+	return m.name
+}
+
+// ParseTypes converts TOML string configuration to TypeMatcher slice.
+func ParseTypes(typeNames []string) ([]TypeMatcher, error) {
```

Comment:
I feel like there might be a less complex approach to mapping between type name strings and their AST types, but that's something we can return to later.

Reply:
I tried using `Reflect` and it is possible to make it work using that, but I liked the matcher pattern better because with `Reflect` we'll still need to implement isEqualTo check. This is because not everything is a "type", some of them are structs and the `BaseType` is a single type. I felt that this was the cleanest API other than rewriting the types as enums and then having to maintain another representation of them.

Reply:
Yes, I agree this is better than `Reflect`. I thought there might be an easier approach based on some `thriftrw-go` functions, but nothing jumped out at me. We can look again later.